jlm/mlir/frontend/MlirToRvsdg.cpp

```diff
@@ -0,0 +1,256 @@
+  JLM_ASSERT(region.getBlocks().size() == 1);
+  return convertBlock(region.front(), rvsdgRegion);
+}
+
+std::unique_ptr<std::vector<jlm::rvsdg::output *>>
+MlirToRvsdg::convertBlock(mlir::Block & block, rvsdg::region & rvsdgRegion)
+{
+  mlir::sortTopologically(&block);
+
+  // Create an RVSDG node for each MLIR operation and store each pair in a
+  // hash map for easy lookup of corresponding RVSDG nodes
+  std::unordered_map<mlir::Operation *, rvsdg::node *> operations;
+  for (mlir::Operation & mlirOp : block.getOperations())
+  {
+    std::vector<const jlm::rvsdg::output *> inputs;
+    for (mlir::Value operand : mlirOp.getOperands())
+    {
+      if (mlir::Operation * producer = operand.getDefiningOp())
+      {
+        size_t index = getOperandIndex(producer, operand);
+        inputs.push_back(operations[producer]->output(index));
+      }
+      else
+      {
+        // If there is no defining op, the Value is necessarily a Block argument.
+        auto blockArg = operand.cast<mlir::BlockArgument>();
+        inputs.push_back(rvsdgRegion.argument(blockArg.getArgNumber()));
+      }
+    }
+
+    if (rvsdg::node * node = convertOperation(mlirOp, rvsdgRegion, inputs))
+    {
+      operations[&mlirOp] = node;
+    }
+  }
+
+  // The results of the region/block are encoded in the terminator operation
+  auto terminator = block.getTerminator();
+  std::unique_ptr<std::vector<jlm::rvsdg::output *>> results =
+      std::make_unique<std::vector<jlm::rvsdg::output *>>();
+  for (mlir::Value operand : terminator->getOperands())
+  {
+    if (mlir::Operation * producer = operand.getDefiningOp())
+    {
+      size_t index = getOperandIndex(producer, operand);
+      results->push_back(operations[producer]->output(index));
+    }
+    else
+    {
+      // If there is no defining op, the Value is necessarily a Block argument.
+      auto blockArg = operand.cast<mlir::BlockArgument>();
+      results->push_back(rvsdgRegion.argument(blockArg.getArgNumber()));
+    }
+  }
+
+  return results;
+}
+
+rvsdg::node *
+MlirToRvsdg::convertOperation(
+    mlir::Operation & mlirOperation,
+    rvsdg::region & rvsdgRegion,
+    std::vector<const rvsdg::output *> & inputs)
+{
+  if (mlirOperation.getName().getStringRef() == mlir::rvsdg::OmegaNode::getOperationName())
+  {
+    convertOmega(mlirOperation, rvsdgRegion);
+    // Omega doesn't have a corresponding RVSDG node so we return nullptr
+    return nullptr;
+  }
+  else if (mlirOperation.getName().getStringRef() == mlir::rvsdg::LambdaNode::getOperationName())
+  {
+    return convertLambda(mlirOperation, rvsdgRegion);
+  }
+  else if (mlirOperation.getName().getStringRef() == mlir::arith::ConstantIntOp::getOperationName())
+  {
+    auto constant = static_cast<mlir::arith::ConstantIntOp>(&mlirOperation);
+    auto type = constant.getType();
+    JLM_ASSERT(type.getTypeID() == mlir::IntegerType::getTypeID());
+    auto * integerType = static_cast<mlir::IntegerType *>(&type);
```

Comment:
Mlir has its "own" cast operators, similar to LLVM. I think you should use those instead of static_cast.

Reply:
Changed to use ::mlir::cast, ::mlir::dyn_cast, and ::mlir::isa